python/lsst/obs/base/defineVisits.py

```diff
@@ -0,0 +1,659 @@
+            ``self.butler.collections``.
+            Can be any of the types supported by the ``collections`` argument
+            to butler construction.
+
+        Returns
+        -------
+        visitRegion : `lsst.sphgeom.Region`
+            Region for the full visit.
+        visitDetectorRegions : `dict` [ `int`, `lsst.sphgeom.Region` ]
+            Dictionary mapping detector ID to the region for that detector.
+            Should include all detectors in the visit.
+        """
+        raise NotImplementedError()
+
+
+class DefineVisitsConfig(Config):
+    groupExposures = GroupExposuresTask.registry.makeField(
+        doc="Algorithm for grouping exposures into visits.",
+        default="one-to-one",
+    )
+    computeVisitRegions = ComputeVisitRegionsTask.registry.makeField(
+        doc="Algorithm from computing visit and visit+detector regions.",
+        default="single-raw-wcs",
+    )
+    ignoreNonScienceExposures = Field(
+        doc=("If True, silently ignore input exposures that do not have "
+             "observation_type=SCIENCE.  If False, raise an exception if one "
+             "encountered."),
+        dtype=bool,
+        optional=False,
+        default=True,
+    )
+
+
+class DefineVisitsTask(Task):
+    """Driver Task for defining visits (and their spatial regions) in Gen3
+    Butler repositories.
+
+    Parameters
+    ----------
+    config : `DefineVisitsConfig`
+        Configuration for the task.
+    butler : `~lsst.daf.butler.Butler`
+        Writeable butler instance.  Will be used to read `raw.wcs` and `camera`
+        datasets and insert/sync dimension data.
+    **kwargs
+        Additional keyword arguments are forwarded to the `lsst.pipe.base.Task`
+        constructor.
+
+    Notes
+    -----
+    Each instance of `DefineVisitsTask` reads from / writes to the same Butler.
+    Each invocation of `DefineVisitsTask.run` processes an independent group of
+    exposures into one or more new vists, all belonging to the same visit
+    system and instrument.
+
+    The actual work of grouping exposures and computing regions is delegated
+    to pluggable subtasks (`GroupExposuresTask` and `ComputeVisitRegionsTask`),
+    respectively.  The defaults are to create one visit for every exposure,
+    and to use exactly one (arbitrary) detector-level raw dataset's WCS along
+    with camera geometry to compute regions for all detectors.  Other
+    implementations can be created and configured for instruments for which
+    these choices are unsuitable (e.g. because visits and exposures are not
+    one-to-one, or because ``raw.wcs`` datasets for different detectors may not
+    be consistent with camera geomery).
+
+    It is not necessary in general to ingest all raws for an exposure before
+    defining a visit that includes the exposure; this depends entirely on the
+    `ComputeVisitRegionTask` subclass used.  For the default configuration,
+    a single raw for each exposure is sufficient.
```

Comment:
Yes and yes.

Unless we can find a way to remove the full-visit region from the visit table, assuming all detectors are included in the region _and_ that the information from one detector is enough to build regions for all of them is the only way we'll be able to make visit definition work in AP when we've got only one detector for one or more of the exposures (i.e. so we can start visit-level tasks as soon as the first raw in the last exposure rolls in).

That's not quite as clean as I'd like yet anyway because I'm still using `insertDimensionData` for the `visit` record itself here; it'll work for that first raw, but any attempt to define visits when the later ones land will fail rather than recognizing that the visit already exists and doing nothing.  That requires whatever is triggering visit definition to know about the order in which different raws land, and I'm guessing that's at least undesirable.  I have an idea for how to make visit definition idempotent, and this discussion reminded me to add a note about that to [DM-24355](https://jira.lsstcorp.org/browse/DM-24355) (which I've now done).